Theory_assignment_two/assignment_2.md

```diff
@@ -0,0 +1,453 @@
+|     Testing occurs concurrently with software   development    |     Testing comes after the building phase       |   |   |   |
+|                                                                |                                                  |   |   |   |
+
+
+
+
+
+
+
+## 7. What is a reduced function used for? 
+
+Reduce() is used for applying a function to an iterable and reducing it to a single cumulative value. It performs function on first two elements and repeats process until 1 value remains.
+
+```python
+#example of a reduced function
+import functools
+
+letters = ["H","E","L","L","O"]
+word = functools.reduce(lambda x, y, :x + y,letters)
+print(word)
+```
+
+## 8. How does merge sort work 
+
+A merge sort is a sorting algorithm that is based on divide and conquer. An array is divided into two. The subarray are then divided again and again until the arrays only have one element each.  Then pairs of one-element arrays are sorted and are merged with other pairs until the initial array is sorted.
+
+```python
+# example of merge sort function taken from https://www.geeksforgeeks.org/merge-sort/
+
+# Python program for implementation of MergeSort
+def mergeSort(arr):
+	if len(arr) > 1:
+
+		# Finding the mid of the array
+		mid = len(arr)//2
+
+		# Dividing the array elements
+		L = arr[:mid]
+
+		# into 2 halves
+		R = arr[mid:]
+
+		# Sorting the first half
+		mergeSort(L)
+
+		# Sorting the second half
+		mergeSort(R)
+
+		i = j = k = 0
+
+		# Copy data to temp arrays L[] and R[]
+		while i < len(L) and j < len(R):
+			if L[i] <= R[j]:
+				arr[k] = L[i]
+				i += 1
+			else:
+				arr[k] = R[j]
+				j += 1
+			k += 1
+
+		# Checking if any element was left
+		while i < len(L):
+			arr[k] = L[i]
+			i += 1
+			k += 1
+
+		while j < len(R):
+			arr[k] = R[j]
+			j += 1
+			k += 1
+
+def printList(arr):
+	for i in range(len(arr)):
+		print(arr[i], end=" ")
+	print()
+
+
+# Driver Code
+if __name__ == '__main__':
+	arr = [12, 11, 13, 5, 6, 7]
+	print("Given array is", end="\n")
+	printList(arr)
+	mergeSort(arr)
+	print("Sorted array is: ", end="\n")
+	printList(arr)
+```
+## 9. Generators - Generator functions allow you to declare a function that behaves like an iterator, i.e. it can be used in a for loop. What is the use case? 
+
+A use case for generators can be for reading very large files instead of reading files line by line.
+
+```python
+
+def csv_reader(file_name):
+    for row in open(file_name, "r"):
+        yield row
+ ```
+
+
+## 10. Decorators - A page for useful (or potentially abusive?) decorator ideas. What is the return type of the decorator? 
+
+A decorator is a function which takes another function as an argument and returns a modified version of it, enhancing its functionality in some way.
+
+Types of decorators include simple decorators, chaining decorators and universal decorators.
+
+Simple decorators – uses a single decorator, function and wrapper
```

Comment:
6/10 The main task of this question is to throw some ideas of how decorators can be potentially used. Some use cases on how decorators can be useful and how it can be abusive (if any), and not just an example of how it works.